backend/src/Edna/Web/API.hs

```diff
@@ -32,3 +44,159 @@ type EdnaAPI =
 
 ednaAPI :: Proxy EdnaAPI
 ednaAPI = Proxy
+
+-- | Endpoints necessary to implement file uploading.
+data FileUploadEndpoints route = FileUploadEndpoints
+  { -- | Parse the file and return its summary for preview.
+    fueParseFile :: route
+      :- "file"
+      :> "parse"
+      :> Summary "Parse the file and return its summary for preview"
+      :> MultipartForm Mem (MultipartData Mem)
+      :> Get '[JSON] FileSummary
+
+  , -- | Upload the file with some methodology and project.
+    fueUploadFile :: route
+      :- "file"
+      :> "upload"
+      :> Summary "Upload the file with some methodology and project"
+      :> ReqBody '[JSON] FileUploadInput
+      :> MultipartForm Mem (MultipartData Mem)
+      :> Post '[JSON] FileSummary
+  } deriving stock (Generic)
+
+type FileUploadAPI = ToServant FileUploadEndpoints AsApi
+
+-- TODO: pagination and sorting are just stubs for now (everywhere).
+-- Most likely we will use @servant-util@ to implement them,
+-- but let's do it later.
+
+-- | Endpoints related to projects.
+data ProjectEndpoints route = ProjectEndpoints
+  { -- | Add a new project.
+    peAddProject :: route
+      :- "project"
+      :> Summary "Add a new project"
+      :> ReqBody '[JSON] Project
+      :> Post '[JSON] (WithId Project)
+
+  , -- | Update an existing project.
+    peEditProject :: route
+      :- "project"
+      :> Summary "Update an existing project"
+      :> Capture "projectId" (SqlId Project)
+      :> ReqBody '[JSON] Project
+      :> Put '[JSON] (WithId Project)
+
+  , -- | Get known projects with optional pagination and sorting
+    peGetProjects :: route
+      :- "projects"
+      :> Summary "Get known projects"
+      :> QueryParam "page" Word
+      :> QueryParam "size" Word
+      :> QueryParam "sortby" StubSortBy
+      :> Get '[JSON] [WithExtra Project ProjectExtra]
+
+  , -- | Get project data by ID
+    peGetProject :: route
+      :- "project"
+      :> Summary "Get project data by ID"
+      :> Capture "projectId" (SqlId Project)
+      :> Get '[JSON] (WithExtra Project ProjectExtra)
+  } deriving stock (Generic)
```

Comment:
Will write here comments which which also relate to the code below.\

```haskell
:> ReqBody '[JSON] Project
:> Post '[JSON] (WithId Project)
```
Why you do not write `WithExtra` here? If you add project from `Library` page you need to know its extra information\

```haskell
      :> Put '[JSON] (WithId Project)
```
Same here, don't understand why we do not want to get extra information\

You can combine edit and add entry-points making id optional\



```haskell
WithExtra Project ProjectExtra
```
So, as for `WithExtra` and `WithId` are not needed. As we want to use MVC model, I propose to just create data types like `ProjectUpdateReq` and okay `SqlId Project` for requests (also I am not sure that `SqlId` is a good naming for web entry points) and `ProjectRep` for responses

Reply:
> Why you do not write WithExtra here? If you add project from Library page you need to know its extra information\

Because this is what @pva701 [told me](https://serokell.slack.com/archives/C01F8AQSHGE/p1615480272097200?thread_ts=1615459098.060600&cid=C01F8AQSHGE). He is the consumer of this data, so I just return what he wants.

> You can combine edit and add entry-points making id optional\

Yes, I [was considering](https://serokell.slack.com/archives/C01F8AQSHGE/p1615461983072300?thread_ts=1615459098.060600&cid=C01F8AQSHGE) this. I don't think it matters, both options sound fine to me.

> So, as for me WithExtra and WithId are not needed. As we want to use MVC model, I propose to just create data types like ProjectUpdateReq and SqlId Project for requests (also I am not sure that SqlId is a good naming for web entry points) and ProjectRep for responses

I already like that I defined `WithExtra` and `WithId`, because now I can quickly change `WithId` to `WithExtra` and vice versa. Or remove ID and Extra completely.
I think `WithId` is useful at least for methodology and project endpoints because there were have `Project` and `TestMethodology` types as submitted by users and we need to return these types with `Id` attached. So instead of defining `ProjectRep` and `MethodologyRep` we can define one `WithId`. For `Compound` and `Target` it's less useful, but still I see a pattern here. Regarding `WithExtra` the situation is similar.

All in all I want to once again emphasize that these types may change in Haskell because currently we have no implementation and probably actual implementation will affect something. Maybe sometimes will be moved to lower levels, maybe not. Hopefully most of these changes will affect only backend, not frontend.

Reply:
> also I am not sure that `SqlId` is a good naming for web entry points

Oh, regarding name, there is again a question of where this type will reside eventually. And also I just couldn't find a good name, I could call it `Id`, but it's too general, so I needed some prefix and couldn't find a better one.

Reply:
Actually one of the reasons why there is no `ProjectRep` is that response is different in different endpoints. In `getProject` we want to return all data (including extra), in `addProject` we just want to return `WithId Project`. It was the case on Thursday when @pva701 told me that. Now he apparently has changed his mind, so the structure of responses becomes more uniform. So it partially defeats the purpose of `WithExtra` at least.

To sum up, while I admit that these types are controversial (they were supposed to be experimental from the beginning), I am not convinced that we should change them _now_. So I propose to merge this PR with these types and revise them later.